File: php/src/LombokAlgoritma/Hash/XxHash64.php
<?php

// LombokAlgoritma — xxHash64 / XXH64 (SPEC §12)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

declare(strict_types=1);

namespace LombokAlgoritma\Hash;

use LombokAlgoritma\Core\U64;

/**
 * XXH64 per the reference specification (doc/xxhash_spec.md "XXH64 Algorithm Description").
 * All arithmetic is wrapping u64 via {@see U64}.
 */
final class XxHash64
{
    private const P1 = -7046029288634856825; // 0x9e3779b185ebca87
    private const P2 = -4417276706812531889; // 0xc2b2ae3d27d4eb4f
    private const P3 = 1609587929392839161; // 0x165667b19e3779f9
    private const P4 = -8796714831421723037; // 0x85ebca77c2b2ae63
    private const P5 = 2870177450012600261; // 0x27d4eb2f165667c5

    /**
     * XXH64 of `$data` with a 64-bit seed (int, decimal string or GMP, reduced mod 2^64), as 16
     * lowercase hex digits.
     */
    public static function hash(string $data, int|string|\GMP $seed = 0): string
    {
        return U64::toHex(self::hashInt($data, $seed));
    }

    /** XXH64 as a u64 bit pattern (PHP int). */
    public static function hashInt(string $data, int|string|\GMP $seed = 0): int
    {
        $s = U64::from($seed);
        $n = strlen($data);
        $i = 0;
        if ($n >= 32) {
            $v1 = U64::add(U64::add($s, self::P1), self::P2);
            $v2 = U64::add($s, self::P2);
            $v3 = $s;
            $v4 = U64::sub($s, self::P1);
            while ($i + 32 <= $n) {
                $v1 = self::round($v1, U64::readLe($data, $i));
                $v2 = self::round($v2, U64::readLe($data, $i + 8));
                $v3 = self::round($v3, U64::readLe($data, $i + 16));
                $v4 = self::round($v4, U64::readLe($data, $i + 24));
                $i += 32;
            }
            $h = U64::add(
                U64::add(U64::rotl($v1, 1), U64::rotl($v2, 7)),
                U64::add(U64::rotl($v3, 12), U64::rotl($v4, 18)),
            );
            $h = self::mergeRound($h, $v1);
            $h = self::mergeRound($h, $v2);
            $h = self::mergeRound($h, $v3);
            $h = self::mergeRound($h, $v4);
        } else {
            $h = U64::add($s, self::P5);
        }
        $h = U64::add($h, $n);
        while ($i + 8 <= $n) {
            $h ^= self::round(0, U64::readLe($data, $i));
            $h = U64::add(U64::mul(U64::rotl($h, 27), self::P1), self::P4);
            $i += 8;
        }
        if ($i + 4 <= $n) {
            $w = ord($data[$i]) | (ord($data[$i + 1]) << 8) | (ord($data[$i + 2]) << 16)
                | (ord($data[$i + 3]) << 24);
            $h ^= U64::mul($w, self::P1);
            $h = U64::add(U64::mul(U64::rotl($h, 23), self::P2), self::P3);
            $i += 4;
        }
        while ($i < $n) {
            $h ^= U64::mul(ord($data[$i]), self::P5);
            $h = U64::mul(U64::rotl($h, 11), self::P1);
            $i++;
        }
        $h ^= U64::shr($h, 33);
        $h = U64::mul($h, self::P2);
        $h ^= U64::shr($h, 29);
        $h = U64::mul($h, self::P3);
        return $h ^ U64::shr($h, 32);
    }

    private static function round(int $acc, int $lane): int
    {
        return U64::mul(U64::rotl(U64::add($acc, U64::mul($lane, self::P2)), 31), self::P1);
    }

    private static function mergeRound(int $acc, int $val): int
    {
        return U64::add(U64::mul($acc ^ self::round(0, $val), self::P1), self::P4);
    }
}

File: rust/lombokalgoritma/src/string/hash/xxhash64.rs
// LombokAlgoritma — xxHash64 (SPEC §12)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

const P1: u64 = 0x9e37_79b1_85eb_ca87;
const P2: u64 = 0xc2b2_ae3d_27d4_eb4f;
const P3: u64 = 0x1656_67b1_9e37_79f9;
const P4: u64 = 0x85eb_ca77_c2b2_ae63;
const P5: u64 = 0x27d4_eb2f_1656_67c5;

fn le64(b: &[u8]) -> u64 {
    let mut w = [0u8; 8];
    w.copy_from_slice(&b[..8]);
    u64::from_le_bytes(w)
}

fn round(acc: u64, lane: u64) -> u64 {
    acc.wrapping_add(lane.wrapping_mul(P2))
        .rotate_left(31)
        .wrapping_mul(P1)
}

fn merge_round(acc: u64, val: u64) -> u64 {
    (acc ^ round(0, val)).wrapping_mul(P1).wrapping_add(P4)
}

/// xxHash64 (XXH64) per the reference specification (`doc/xxhash_spec.md`).
pub fn xxhash64(data: &[u8], seed: u64) -> u64 {
    let n = data.len();
    let mut i = 0;
    let mut h = if n >= 32 {
        let mut v = [
            seed.wrapping_add(P1).wrapping_add(P2),
            seed.wrapping_add(P2),
            seed,
            seed.wrapping_sub(P1),
        ];
        while i + 32 <= n {
            for (lane, acc) in v.iter_mut().enumerate() {
                *acc = round(*acc, le64(&data[i + 8 * lane..]));
            }
            i += 32;
        }
        let mut h = v[0]
            .rotate_left(1)
            .wrapping_add(v[1].rotate_left(7))
            .wrapping_add(v[2].rotate_left(12))
            .wrapping_add(v[3].rotate_left(18));
        for lane in v {
            h = merge_round(h, lane);
        }
        h
    } else {
        seed.wrapping_add(P5)
    };
    h = h.wrapping_add(n as u64);
    while i + 8 <= n {
        h ^= round(0, le64(&data[i..]));
        h = h.rotate_left(27).wrapping_mul(P1).wrapping_add(P4);
        i += 8;
    }
    if i + 4 <= n {
        let w = u64::from(u32::from_le_bytes([
            data[i],
            data[i + 1],
            data[i + 2],
            data[i + 3],
        ]));
        h ^= w.wrapping_mul(P1);
        h = h.rotate_left(23).wrapping_mul(P2).wrapping_add(P3);
        i += 4;
    }
    while i < n {
        h ^= u64::from(data[i]).wrapping_mul(P5);
        h = h.rotate_left(11).wrapping_mul(P1);
        i += 1;
    }
    h ^= h >> 33;
    h = h.wrapping_mul(P2);
    h ^= h >> 29;
    h = h.wrapping_mul(P3);
    h ^ (h >> 32)
}
